File: rss_worker/summariser.py
"""Claude API summarisation for RSS articles."""

import json
import logging
import os
import time

import anthropic

logger = logging.getLogger(__name__)

# Reuse a single Anthropic client across all summarise() calls to avoid
# re-creating the HTTP client (and its connection pool) per article.
_anthropic_client: anthropic.Anthropic | None = None

# Transient errors worth retrying
_RETRYABLE = (
    anthropic.APITimeoutError,
    anthropic.RateLimitError,
    anthropic.InternalServerError,
    anthropic.APIConnectionError,
)

_MAX_RETRIES = 2
_RETRY_DELAY = 3


def _get_client() -> anthropic.Anthropic | None:
    """Return a cached Anthropic client, or None if no valid API key."""
    global _anthropic_client
    if _anthropic_client is not None:
        return _anthropic_client
    api_key = os.getenv("ANTHROPIC_API_KEY", "")
    if not api_key or api_key == "your_key_here":
        return None
    _anthropic_client = anthropic.Anthropic(api_key=api_key)
    return _anthropic_client


def _is_refusal(summary: str) -> bool:
    """Detect when the model refuses to summarise (e.g. can't access URLs)."""
    lower = summary.lower()
    refusal_phrases = [
        "i don't have access",
        "i do not have access",
        "i'm unable to",
        "i am unable to",
        "i cannot access",
        "i can't access",
        "i can't browse",
        "i cannot browse",
        "i'm not able to",
        "i am not able to",
    ]
    return any(phrase in lower for phrase in refusal_phrases)


def summarise(title: str, url: str, content_text: str) -> str | None:
    """Summarise an article using Claude Haiku.

    Returns None if summarisation fails or the model refuses (e.g. cannot
    access the URL).  Falls back to truncation only when the API is
    unavailable or hits transient errors AND valid content was provided.

    Args:
        title: Article title.
        url: Article URL.
        content_text: Plain text content of the article.

    Returns:
        A 2-3 sentence summary, or None if the article could not be
        meaningfully summarised.
    """
    client = _get_client()
    if client is None:
        logger.warning("No valid ANTHROPIC_API_KEY set, falling back to truncation")
        return _fallback_summary(title, content_text)

    last_exc: Exception | None = None
    for attempt in range(1, _MAX_RETRIES + 1):
        try:
            message = client.messages.create(
                model="claude-haiku-4-5-20251001",
                max_tokens=256,
                messages=[
                    {
                        "role": "user",
                        "content": (
                            "Summarise the following article in 2-3 sentences, "
                            "focusing on what a developer might find actionable or useful. "
                            "Include the main technology or concept. Be concise.\n\n"
                            f"Title: {title}\n"
                            f"URL: {url}\n\n"
                            f"{content_text[:3000]}"
                        ),
                    }
                ],
            )
            summary = message.content[0].text.strip()
            if _is_refusal(summary):
                logger.warning(
                    "Model refused to summarise '%s' (%s) — skipping",
                    title, url,
                )
                return None
            logger.debug("Summarised: %s", title)
            return summary
        except _RETRYABLE as exc:
            last_exc = exc
            logger.warning(
                "Summarisation attempt %d/%d for '%s' failed: %s: %s",
                attempt, _MAX_RETRIES, title, type(exc).__name__, exc,
            )
            if attempt < _MAX_RETRIES:
                time.sleep(_RETRY_DELAY * attempt)
        except Exception as exc:
            logger.error(
                "Summarisation failed for '%s': %s: %s", title, type(exc).__name__, exc,
            )
            return _fallback_summary(title, content_text)

    logger.error(
        "Summarisation failed after %d retries for '%s': %s: %s",
        _MAX_RETRIES, title, type(last_exc).__name__, last_exc,
    )
    return _fallback_summary(title, content_text)


def _fallback_summary(title: str, content_text: str) -> str:
    """Truncation fallback when API is unavailable."""
    preview = content_text[:800].strip()
    if len(content_text) > 800:
        preview += "..."
    return f"{title}. {preview}"


_EXTRACT_PROMPT = (
    "You are analysing an article to extract individual news items, "
    "announcements, or updates. The article may cover a single topic "
    "or be a newsletter/digest containing many distinct items.\n\n"
    "For EACH distinct item, project, announcement, or piece of news, extract:\n"
    "- title: A short descriptive title (max 10 words)\n"
    "- who: Who or what is this about? (person, project, organisation, technology)\n"
    "- what: What happened, what does it do, or what was announced?\n"
    "- why: Why does this matter? Why should the reader care?\n\n"
    "Return ONLY a valid JSON array of objects. No markdown fences, no explanation.\n"
    "If the article covers a single topic, return an array with one object.\n"
    "Skip promotional content, sponsor mentions, and calls-to-action.\n"
    "Skip items where you cannot determine meaningful who/what/why.\n\n"
)


def extract_items(
    title: str, url: str, content_text: str,
) -> list[dict[str, str]] | None:
    """Extract individual items from a newsletter or multi-topic article.

    Uses Claude to identify each distinct news item, project, or announcement
    and return structured who/what/why for each.

    Args:
        title: Article title.
        url: Article URL.
        content_text: Plain text content of the article.

    Returns:
        List of dicts with title/who/what/why keys, or None on failure.
    """
    client = _get_client()
    if client is None:
        logger.warning("No valid ANTHROPIC_API_KEY — cannot extract items")
        return None

    last_exc: Exception | None = None
    for attempt in range(1, _MAX_RETRIES + 1):
        try:
            message = client.messages.create(
                model="claude-haiku-4-5-20251001",
                max_tokens=4096,
                messages=[
                    {
                        "role": "user",
                        "content": (
                            f"{_EXTRACT_PROMPT}"
                            f"Article title: {title}\n"
                            f"Article URL: {url}\n\n"
                            f"{content_text[:12000]}"
                        ),
                    }
                ],
            )

            response_text = message.content[0].text.strip()

            if _is_refusal(response_text):
                logger.warning(
                    "Model refused to extract items from '%s' (%s) — skipping",
                    title, url,
                )
                return None

            # Strip markdown fences if the model wrapped the JSON
            if response_text.startswith("```"):
                response_text = response_text.split("\n", 1)[1]
                response_text = response_text.rsplit("```", 1)[0].strip()

            items = json.loads(response_text)
            if not isinstance(items, list):
                logger.warning(
                    "Expected JSON array from extraction of '%s', got %s",
                    title, type(items).__name__,
                )
                return None

            valid: list[dict[str, str]] = []
            required = ("title", "who", "what", "why")
            for item in items:
                if isinstance(item, dict) and all(k in item for k in required):
                    valid.append({k: str(item[k]).strip() for k in required})

            if not valid:
                logger.warning("No valid items extracted from '%s'", title)
                return None

            logger.info("Extracted %d items from '%s'", len(valid), title)
            return valid

        except json.JSONDecodeError as exc:
            logger.warning(
                "Failed to parse JSON from extraction of '%s': %s", title, exc,
            )
            return None
        except _RETRYABLE as exc:
            last_exc = exc
            logger.warning(
                "Extraction attempt %d/%d for '%s' failed: %s: %s",
                attempt, _MAX_RETRIES, title, type(exc).__name__, exc,
            )
            if attempt < _MAX_RETRIES:
                time.sleep(_RETRY_DELAY * attempt)
        except Exception as exc:
            logger.error(
                "Extraction failed for '%s': %s: %s",
                title, type(exc).__name__, exc,
            )
            return None

    logger.error(
        "Extraction failed after %d retries for '%s': %s: %s",
        _MAX_RETRIES, title, type(last_exc).__name__, last_exc,
    )
    return None

File: crates/omnimem-rss/src/summariser.rs
//! Article summaries and digest item extraction (`rss_worker/summariser.py`).
//!
//! 6.x wrapped each call in its own two-attempt retry on top of the SDK's;
//! the `omnimem-llm` client already retries the same failures, so a failed
//! call here goes straight to the fallback.

use omnimem_core::LanguageModel;
use omnimem_engine::pyfmt::take_chars;
use omnimem_engine::skills::py_str;
use serde_json::Value;
use tracing::{debug, error, info, warn};

/// 6.x hard-coded the model for both calls.
pub const MODEL: &str = "claude-haiku-4-5-20251001";

const REFUSAL_PHRASES: [&str; 10] = [
    "i don't have access",
    "i do not have access",
    "i'm unable to",
    "i am unable to",
    "i cannot access",
    "i can't access",
    "i can't browse",
    "i cannot browse",
    "i'm not able to",
    "i am not able to",
];

const EXTRACT_PROMPT: &str = "You are analysing an article to extract individual news items, \
announcements, or updates. The article may cover a single topic \
or be a newsletter/digest containing many distinct items.\n\n\
For EACH distinct item, project, announcement, or piece of news, extract:\n\
- title: A short descriptive title (max 10 words)\n\
- who: Who or what is this about? (person, project, organisation, technology)\n\
- what: What happened, what does it do, or what was announced?\n\
- why: Why does this matter? Why should the reader care?\n\n\
Return ONLY a valid JSON array of objects. No markdown fences, no explanation.\n\
If the article covers a single topic, return an array with one object.\n\
Skip promotional content, sponsor mentions, and calls-to-action.\n\
Skip items where you cannot determine meaningful who/what/why.\n\n";

/// One item pulled out of a digest article.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DigestItem {
    pub title: String,
    pub who: String,
    pub what: String,
    pub why: String,
}

/// The model declined, typically because it was asked about a URL.
pub fn is_refusal(text: &str) -> bool {
    let lower = text.to_lowercase();
    REFUSAL_PHRASES.iter().any(|p| lower.contains(p))
}

/// The truncation stand-in when no summary can be had.
pub fn fallback_summary(title: &str, content: &str) -> String {
    let mut preview = take_chars(content, 800).trim().to_owned();
    if content.chars().count() > 800 {
        preview.push_str("...");
    }
    format!("{title}. {preview}")
}

/// A two or three sentence summary, the truncation fallback when there is no
/// model or the call fails, or `None` when the model refuses.
pub fn summarise(
    llm: Option<&dyn LanguageModel>,
    title: &str,
    url: &str,
    content: &str,
) -> Option<String> {
    let Some(llm) = llm else {
        warn!("No valid ANTHROPIC_API_KEY set, falling back to truncation");
        return Some(fallback_summary(title, content));
    };
    let prompt = format!(
        "Summarise the following article in 2-3 sentences, focusing on what a developer might find \
         actionable or useful. Include the main technology or concept. Be concise.\n\n\
         Title: {title}\nURL: {url}\n\n{}",
        take_chars(content, 3000)
    );
    match llm.complete(MODEL, &prompt, 256) {
        Ok(reply) => {
            let summary = reply.trim();
            if is_refusal(summary) {
                warn!(title, url, "model refused to summarise, skipping");
                return None;
            }
            debug!(title, "summarised");
            Some(summary.to_owned())
        }
        Err(e) => {
            error!(title, error = %e, "summarisation failed");
            Some(fallback_summary(title, content))
        }
    }
}

/// Each distinct item in a newsletter or multi-topic article, or `None` when
/// there is no model, the call fails, the model refuses or nothing is usable.
pub fn extract_items(
    llm: Option<&dyn LanguageModel>,
    title: &str,
    url: &str,
    content: &str,
) -> Option<Vec<DigestItem>> {
    let Some(llm) = llm else {
        warn!("No valid ANTHROPIC_API_KEY, cannot extract items");
        return None;
    };
    let prompt = format!(
        "{EXTRACT_PROMPT}Article title: {title}\nArticle URL: {url}\n\n{}",
        take_chars(content, 12_000)
    );
    let reply = match llm.complete(MODEL, &prompt, 4096) {
        Ok(reply) => reply,
        Err(e) => {
            error!(title, error = %e, "extraction failed");
            return None;
        }
    };
    let mut text = reply.trim().to_owned();
    if is_refusal(&text) {
        warn!(title, url, "model refused to extract items, skipping");
        return None;
    }
    if text.starts_with("```") {
        let Some((_, rest)) = text.split_once('\n') else {
            error!(title, "extraction failed: unterminated code fence");
            return None;
        };
        text = rest
            .rfind("```")
            .map_or(rest, |i| &rest[..i])
            .trim()
            .to_owned();
    }
    let items = match serde_json::from_str::<Value>(&text) {
        Ok(Value::Array(items)) => items,
        Ok(_) => {
            warn!(title, "expected a JSON array from extraction");
            return None;
        }
        Err(e) => {
            warn!(title, error = %e, "failed to parse JSON from extraction");
            return None;
        }
    };
    let valid: Vec<DigestItem> = items
        .iter()
        .filter_map(Value::as_object)
        .filter(|item| {
            ["title", "who", "what", "why"]
                .iter()
                .all(|k| item.contains_key(*k))
        })
        .map(|item| {
            let field = |k: &str| py_str(&item[k]).trim().to_owned();
            DigestItem {
                title: field("title"),
                who: field("who"),
                what: field("what"),
                why: field("why"),
            }
        })
        .collect();
    if valid.is_empty() {
        warn!(title, "no valid items extracted");
        return None;
    }
    info!(title, items = valid.len(), "extracted digest items");
    Some(valid)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn refusals_are_spotted_case_insensitively() {
        assert!(is_refusal("Sorry, I'm Unable To open links."));
        assert!(!is_refusal("Rust 1.96 stabilises let chains."));
    }

    #[test]
    fn the_fallback_truncates_at_800_characters() {
        let long = "é".repeat(900);
        let summary = fallback_summary("T", &long);
        assert_eq!(summary.chars().count(), "T. ".len() + 800 + 3);
        assert!(summary.ends_with("..."));
        assert_eq!(fallback_summary("T", "  short  "), "T. short");
    }
}
